recipe/generate_llvm.py

```diff
@@ -0,0 +1,371 @@
+    linkopts_cache[SYMBOL][name] = set(linkopts)
+
+    # Remove static libraries of direct dependencies
+    converted_deps = {f"-l{SYMBOL}{dep[1:]}" for dep in deps}
+    linkopts = [opt for opt in linkopts if opt not in converted_deps]
+
+    # Remove static libraries of indirect dependencies via linkopts_cache
+    provided_deps = set()
+    for dep in deps:
+        provided_deps |= linkopts_cache[SYMBOL].get(dep[1:], set())
+    linkopts = [opt for opt in linkopts if opt not in provided_deps]
+
+    return linkopts
+
+def do_make_linkopts(name):
+    fullname = SYMBOL + name
+    # Fix the caseing of the library to match the actual file names.
+    if fullname.lower() in case_lookup:
+        fullname = case_lookup[fullname.lower()]
+
+    # There are some difference in the names used in Bazel and the actual static libraries.
+    # This lookup here is by done by simply guessing them given
+    # the linkage errors that popped up.
+    if fullname == "MLIRAnalysis":
+        return add_link_dependencies(["MLIRPresburger", "MLIRAnalysis", "MLIRLoopAnalysis"])
+    if fullname == "MLIRCallOpInterfaces":
+        return add_link_dependencies(["MLIRCallInterfaces"])
+    if fullname == "MLIRShapeTransforms":
+        return add_link_dependencies(["MLIRShape", "MLIRShapeOpsTransforms"])
+
+    libname = f"lib{fullname}.a"
+    upper_match = re.search("(.*)([A-Z][a-z]*)$", name)
+    if (PREFIX / "lib" / libname).exists():
+        return add_link_dependencies([f"{fullname}"])
+    elif (PREFIX / "lib" / libname.replace("Dialect", "IR")).exists():
+        return [f"-l{fullname.replace('Dialect', 'IR')}"]
+    elif upper_match is not None:
+        name = upper_match.group(1)
+        if len(name) > 1:
+            return do_make_linkopts(name)
+    return []
+
+
+libdir = Path(os.environ["PREFIX"]) / "lib"
+
+output = """
+exports_files([\"LICENSE.TXT\"])
+"""
+
+r = requests.get(f"https://raw.githubusercontent.com/llvm/llvm-project/{llvm_commit}/llvm/LICENSE.txt")
+Path("llvm-project/llvm/LICENSE.TXT").write_text(r.text)
+
+r = requests.get(f"https://raw.githubusercontent.com/llvm/llvm-project/{llvm_commit}/utils/bazel/llvm-project-overlay/llvm/BUILD.bazel")
+
+LLVM_LIBS = []
+
+llvm_libs = libdir.glob("libLLVM*.a")
+llvm_build = "exports_files([\"LICENSE.TXT\"])"
+for lib in llvm_libs:
+    basename = lib.name[7:-2]
+    LLVM_LIBS.append(basename)
+
+sh_binary = py_binary = binary_alias = filegroup = gentbl = cc_binary = template_rule = genrule = enum_targets_gen = package = load = empty
+llvm_config_defines = []
+# See https://github.com/tensorflow/tensorflow/blob/3f878cff5b698b82eea85db2b60d65a2e320850e/third_party/llvm/setup.bzl#L6
+llvm_targets = [
+    "AArch64",
+    "AMDGPU",
+    "ARM",
+    "NVPTX",
+    "PowerPC",
+    "RISCV",
+    "SystemZ",
+    "X86",
+]
+
+def select(*args, **kwargs):
+    return []
+
+# Load the BUILD file and run it with the above defined custom functions
+exec(r.text)
```

Comment:
Not a big fan of the distance between creating `r` and using it, especially given it's very short name.

Can we do `upstream_build_bazel = r.text` after the `r=` above and then 
```suggestion
# Run the upstream BUILD file with the above defined custom functions
exec(upstream_build_bazel)
```
here?